src/term/check/exhaustiveness.rs

```diff
@@ -0,0 +1,463 @@
+  matches!(pat, RulePat::Var(..))
+}
+
+fn is_wildcard_row(row: &Row) -> bool {
+  row.first().is_some_and(is_wildcard)
+}
+
+fn is_wildcard_matrix(matrix: &Matrix) -> bool {
+  matrix.iter().all(is_wildcard_row)
+}
+
+fn useful(ctx: &Ctx, problem: &mut Problem) -> bool {
+  match problem {
+    // if the matrix has 0 columns and 0 rows the case is useful
+    // to the matrix
+    problem if problem.is_empty() => true,
+
+    // if the matrix has 0 columns but more than 0 rows the case is useless
+    // to the matrix
+    problem if problem.is_complete() => false,
+
+    problem => {
+      // take the types of the patterns
+      if let Some((name, args)) = problem.type_name() {
+        match &problem.case[0] {
+          // if it is a wildcard/var
+          RulePat::Var(..) => {
+            if is_wildcard_matrix(&problem.matrix) {
+              // if all patterns of the first column are wildcards then
+              // get the default matrix and check again
+              useful(ctx, &mut default_matrix(problem))
+            } else {
+              let mut cons: Option<Vec<String>> =
+                problem.matrix.iter().map(|row| get_pat_ctr_name(row[0].clone())).collect();
+              if cons.is_none() {
+                cons = Some(vec![]);
+              }
+              if let Some(names) = cons {
+                match is_sig_complete(ctx, name.0.clone(), names) {
+                  // if the signature is complete then split based on the type
+                  Completeness::Complete(_) => split(ctx, &name, args, problem),
+                  // if the signature of the first pattern is incomplete then get the default matrix
+                  // and check again
+                  Completeness::Incomplete(_) => useful(ctx, &mut default_matrix(problem)),
+                }
+              } else {
+                false
+              }
+            }
+          }
+          // if it is a constructor we specialize this constructor case
+          RulePat::Ctr(nam_, ..) => ctx.specialize_ctr(&nam_, args, &problem),
+        }
+      } else {
+        false
+      }
+    }
+  }
+}
+
+impl Book {
+  pub fn check_exhaustiveness(&self) -> anyhow::Result<()> {
+    let typed_defs = self.typed_defs()?;
+
+    for def in self.defs.values() {
+      if let Some(def_types) = typed_defs.get(&def.def_id) {
+        // get the type of each argument
+        let types = def_types
+          .into_iter()
+          .map(|t| match t {
+            Type::Any => todo!(),
+            Type::Adt(name) => RulePat::Ctr(name.clone(), vec![]),
+          })
+          .collect::<Vec<RulePat>>();
+
+        let matrix = def.get_matrix();
+        let rule_arity = def.rules[0].arity();
+        // this is the default case to check if a definition is exhaustive
+        let case = vec![wildcard(); rule_arity];
+
+        // ctx_cons, ex: [Ctr("T"), Ctr("F")]
+        let mut ctx_cons = HashMap::<String, Vec<RulePat>>::new();
```

Comment:
Both `ctx_cons` and `ctx_types` seem to be just different projections of already existing data. Why not just use `self.adts` and `self.ctrs` directly?

Reply:
I could not use `self.ctrs` because I need the constructor's arity